### DISCUSSION TITLE
Prefetch of WADO ，outcomeMessage：Fetched 0 objects by WADO Exporter WADO-JSON

### DISCUSSION THREAD

**Discussion by bird-bro** [dcm4che/dcm4chee-arc-light#3248] [Q&A] [answered]
Hi, there,
I try to use Prefetch of WADO ，
https://github.com/dcm4che/dcm4chee-arc-light/wiki/Pre-fetch-of-WADO-objects
https://github.com/dcm4che/dcm4chee-arc-light/wiki/Prefetch-WADO-Metadata
But it can't run normally, 
Server Log:Failed to find Web Application for WADO request invocation : Unknown WebApplication，
Archive UI：Monitoring--Export，Status：COMPLETED ，outcomeMessage：	Fetched 0 objects by WADO Exporter WADO-JSON
I'm not sure if there is a configuration error. 
Archive Version 5.23.3

Thanks in advance.

Attach error log and Archive configuration。

![微信图片_20210615162216](https://user-images.githubusercontent.com/39079720/122019003-f372ca00-cdf5-11eb-9812-75514050adbb.png)

![微信截图_20210615162351](https://user-images.githubusercontent.com/39079720/122019151-17cea680-cdf6-11eb-9884-40c9780d1ae5.png)

Using Archive UI：
Web Applications：
![8](https://user-images.githubusercontent.com/39079720/122024136-ba892400-cdfa-11eb-8669-c72b9e894c44.png)

Attribute Set：
![6](https://user-images.githubusercontent.com/39079720/122020194-24073380-cdf7-11eb-929e-9fd151ec4072.png)

Storage Descriptor：
![1](https://user-images.githubusercontent.com/39079720/122020897-bf000d80-cdf7-11eb-9ef9-f62461c97656.png)
![2](https://user-images.githubusercontent.com/39079720/122020906-c1626780-cdf7-11eb-9ffb-ceed59ab8f15.png)

Exporter Descriptor：
![3](https://user-images.githubusercontent.com/39079720/122020966-d3440a80-cdf7-11eb-8ee4-91d9dfa738df.png)
![4](https://user-images.githubusercontent.com/39079720/122020974-d4753780-cdf7-11eb-94b4-c7f79decf209.png)

Export Rule：
![5](https://user-images.githubusercontent.com/39079720/122021015-dccd7280-cdf7-11eb-8a69-abf89c3dcef3.png)

**Comment by vrindanayak** [comment, marked as answer]
Which archive version are you running? The Wado Exporter was redesigned as part of this [issue](https://github.com/dcm4che/dcm4chee-arc-light/issues/3069). The wiki provides configurations for both pre/post 5.23.2 archive version. There was a minor error w.r.t archive version in the wiki which I've corrected. Please refer the wiki again and see the difference for the following

![image](https://user-images.githubusercontent.com/15785031/122174976-8574e400-ce83-11eb-94f4-40616fdf09ab.png)

In addition to the above, your web application configuration is incorrect. You may reuse the existing `DCM4CHEE` web application (which already contains the correct _Web Service Path_ i.e. `/dcm4chee-arc/aets/DCM4CHEE/rs` and _Web Service Classes_ i.e. `WADO_RS`); you would just need to additionally configure a [http connection to be referenced in the web application](https://github.com/dcm4che/dcm4chee-arc-light/wiki/Pre-fetch-of-WADO-objects#use-cases-with-configuration) with `HTTP` as _Protocol_ on `dcm4chee-arc` device and reference this in the `DCM4CHEE` web application's _Network Connection Reference_. 

Also, you don't need an [Attribute Set]() configuration for pre-fetching metadata using Wado Exporter.

**Reply by bird-bro** [reply]
Hi, vrindanayak
  I really appreciate your help！
 After modifying the configuration according to your prompts, it can run successfully on the archive version 5.23.3。

**Comment by bird-bro** [comment]
Hi, vrindanayak
Sorry, it was foun during the test that the metadata obtained was not generated by prefetching.
the json file has been produce and correctly stored in the specified location.

  debug：if (location.getObjectType() == Location.ObjectType.METADATA) == false.
![微信截图_20210621223331](https://user-images.githubusercontent.com/39079720/122781478-6ee6e680-d2e2-11eb-968d-db75ca048675.png)



Attach Archive configuration

Storage Descriptor:
![112](https://user-images.githubusercontent.com/39079720/122781568-87570100-d2e2-11eb-9843-cb9b96d415ca.png)
![113](https://user-images.githubusercontent.com/39079720/122781628-98077700-d2e2-11eb-8d8c-18528e278245.png)

Exporter Descriptor:
![114](https://user-images.githubusercontent.com/39079720/122781654-9e95ee80-d2e2-11eb-8abe-e1e89574f4cf.png)
![115](https://user-images.githubusercontent.com/39079720/122781672-a2c20c00-d2e2-11eb-804a-c1f8002d7256.png)

**Reply by vrindanayak** [reply]
> debug：if (location.getObjectType() == Location.ObjectType.METADATA) == false.

This line is relevant only if you have additionally configured to [store](https://github.com/dcm4che/dcm4chee-arc-light/blob/master/dcm4chee-arc-assembly/src/main/resources/ldap/sample-config.ldif#L3476) [instance metadata](https://github.com/dcm4che/dcm4chee-arc-light/blob/master/dcm4chee-arc-assembly/src/main/resources/ldap/sample-config.ldif#L3449-L3454), whereby the retrieve service first checks if there is such file it loads the attributes from the file, if not then the attributes are read out from DICOM file. So the pre-fetch using wado to store metadata to configured storage descriptor should have worked. What do you see in the `outcomeMessage` / `errorMessage` of the corresponding export tasks in `Monitoring`?

![image](https://user-images.githubusercontent.com/15785031/122913274-890fdb80-d359-11eb-8fff-9ee5a129e1ed.png)

**Reply by bird-bro** [reply]
outcomeMessage：Fetched 1 objects by WADO Exporter WADO-JSON
 I configured to prefetch the metadata of the series,
 using woda (DCM4CHEE/rs/studies/[]series/[]/metadata), 
 I debug the source code , but can't find the place where  the prefetched  json file of the series  is read  into the stream.

![66](https://user-images.githubusercontent.com/39079720/122920304-d315b300-d393-11eb-80ff-7b55fc354292.png)

**Reply by vrindanayak** [reply]
> outcomeMessage：Fetched 1 objects by WADO Exporter WADO-JSON

This indicates that metadata was successfully prefetched into the configured storage. Check the files at `/opt/f1/storage/wado/{now,date,yyyy/MM/dd}/{StudyIUID/{SeriesIUID-hash}.json`

**Reply by bird-bro** [reply]
Yes，I did  find  that  json file generated successfully  in the specified  location . My question is : when I  retrive the metadata with wado request from client browser,   I am not sure waht I get is that prefetched json file or  the one generated on the fly.   
I have tried to debug the source code to see what happened, but unfortunately did not find  the code lines.Simply speaking ,  how can I verify what I get from client browser is read from that prefetched json ?  
  Any help will be appreciated.

**Reply by vrindanayak** [reply]
Not fully sure what exactly you are trying to achieve. 
- If you are planning to build a cache server between client and archive, you don't need to configure a storage for pre-fetching objects. Instead, your cache server (with a separate device configuration and web application) can be configured in the exporter to receive WADO requests which then your cache server forwards back to archive to get the objects pre-fetched and store it in your cache which can be used later by the client for faster retrieval. See [Prefetch of WADO objects - Use Cases](https://github.com/dcm4che/dcm4chee-arc-light/wiki/Pre-fetch-of-WADO-objects#use-cases-with-configuration). 
- Prefetching to a storage was provided as a use case for applications where copies of images / metadata are to be used for some processing by some third party applications. 
- With archive version 5.24.0, [prefetching QIDO responses](https://github.com/dcm4che/dcm4chee-arc-light/issues/3215) shall be possible as well. - [Prefetch QIDO Responses wiki](https://github.com/dcm4che/dcm4chee-arc-light/wiki/Prefetch-QIDO-Responses)
- As previously mentioned, independent of prefetching features, you may configure [storing](https://github.com/dcm4che/dcm4chee-arc-light/blob/master/dcm4chee-arc-assembly/src/main/resources/ldap/sample-config.ldif#L3476) [instance metadata](https://github.com/dcm4che/dcm4chee-arc-light/blob/master/dcm4chee-arc-assembly/src/main/resources/ldap/sample-config.ldif#L3449-L3454) when studies are stored to archive. On subsequent wado requests to archive either using browser or prefetch wado exporters or any other client, the retrieve service shall read the metadata from the instance metadata json files (instead of reading from DICOM files) and send back in the response. In this case, the following 
   > if (location.getObjectType() == Location.ObjectType.METADATA) {
 
  would return `true` for location record of an instance in database which has `object_type` as `METADATA`. 
  Note : Configuring instance metadata storage creates two location records for a single instance - one with `object_type` as `DICOM` and second with `object_type` as `METADATA`

**Reply by bird-bro** [reply]
Hi, vrindanayak
Here is our case:
We have a  dicom  viewer.  Before loading all of the images in one study, we have to  build the study-series-instance tree-like data structure in the viewer. This data structure’s building relies on the study-level metadata retrieved by WADO from dcm4chee. To improve the user experience, it is necessary to get that metadata as fast as possible.
In wiki, this seems what we are looking for:
>1. Pre-fetch the image and/or metadata on some cache server for faster access. In such use cases it is required that the user >configure a separate device and configure HTTP Protocol specific connections on it which are then used for referencing in Web >Application(s) to be configured on this new created device used as a cache device.

But  we don’t know how to configure it.  Any detailed instructions will be helpful.

**Reply by vrindanayak** [reply]
You may build a simple webhook callback application which invokes archive's QIDO / WADO service(s) to get the attributes information. To do this : 
- ensure your application can be reached at `http[s]://<your-webhook-host>:<webhook-port>` and is capable of invoking one (or more) of the following archive's services 
  - For QIDO Prefetch - **Recommended** - Response time of QIDO services is way faster than WADO metadata services when you consider getting metadata for very large studies : 
     - [Search Instances of Series of Study](https://petstore.swagger.io/index.html?url=https://raw.githubusercontent.com/dcm4che/dcm4chee-arc-light/master/dcm4chee-arc-ui2/src/swagger/openapi.json#/QIDO-RS/SearchForStudySeriesInstances) - **Recommended**
     - [Search Instances of Study](https://petstore.swagger.io/index.html?url=https://raw.githubusercontent.com/dcm4che/dcm4chee-arc-light/master/dcm4chee-arc-ui2/src/swagger/openapi.json#/QIDO-RS/SearchForStudyInstances) - Not recommended (as export gets triggered for the whole study - each time any new series / instance is received by archive)
     - [Search Series of Study](https://petstore.swagger.io/index.html?url=https://raw.githubusercontent.com/dcm4che/dcm4chee-arc-light/master/dcm4chee-arc-ui2/src/swagger/openapi.json#/QIDO-RS/SearchForStudySeries) - Not recommended (as export gets triggered for the whole study - each time any new series / instance is received by archive)
  - For WADO Prefetch - **Optional**
     - [Retrieve Series Metadata](https://petstore.swagger.io/index.html?url=https://raw.githubusercontent.com/dcm4che/dcm4chee-arc-light/master/dcm4chee-arc-ui2/src/swagger/openapi.json#/WADO-RS/RetrieveSeriesMetadata) - Lesser preference compared to [Search Instances of Series of Study](https://petstore.swagger.io/index.html?url=https://raw.githubusercontent.com/dcm4che/dcm4chee-arc-light/master/dcm4chee-arc-ui2/src/swagger/openapi.json#/QIDO-RS/SearchForStudySeriesInstances)
     - [Retrieve Study Metadata](https://petstore.swagger.io/index.html?url=https://raw.githubusercontent.com/dcm4che/dcm4chee-arc-light/master/dcm4chee-arc-ui2/src/swagger/openapi.json#/WADO-RS/RetrieveStudyMetadata) - Not recommended (as export gets triggered for the whole study - each time any new series / instance is received by archive)
- Next, configure a separate device (eg. `cache`) in archive configuration with a network connection with `<your-webhook-host>` and `<webhook-port>`. Configure _HTTP_ as _Protocol_ for this network connection.
- In this `cache` device, configure a _Web Application_ (`Devices -> dcm4chee-arc -> Extensions -> Device Extension -> Child Objects -> Web Application`). 
   - Configure _Web Service Path_ as _/dcm4chee-arc/aets/<ARCHIVE_AET>/rs_ (default _<ARCHIVE_AET>_ = _DCM4CHEE_)
   - Select the _HTTP_ network connection in _Web Application Network Connections_
   - Configure _QIDO_RS_ (for QIDO Prefetch) / _WADO_RS_ (for WADO Prefetch) as _Web Service Class_. 
- Follow [Prefetch QIDO responses](https://github.com/dcm4che/dcm4chee-arc-light/wiki/Prefetch-QIDO-Responses) for export rule and exporter configurations (ignore storage configurations), specifying _wado:your-web-app-name_ in the _URI_ field in exporter.
    ````
    dn: dcmExporterID=QIDO-JSON-InstancesOfSeries,dicomDeviceName=dcm4chee-arc,cn=Devices,cn=DICOM Configuration,dc=dcm4che,dc=org
    dcmExporterID: QIDO-JSON-InstancesOfSeries
    dcmQueueName: Export2
    objectClass: dcmExporter
    dicomAETitle: DCM4CHEE
    dcmURI: wado:DCM4CHEE
    dcmProperty: QidoService=/studies/[0]/series/[1]/instances
    dcmProperty: Accept=application/json
    dicomDescription: Export to QIDO-JSON-InstancesOfSeries

    dn: cn=Forward to QIDO-JSON-InstancesOfSeries,dicomDeviceName=dcm4chee-arc,cn=Devices,cn=DICOM Configuration,dc=dcm4che,dc=org
    dcmExporterID: QIDO-JSON-InstancesOfSeries
    dcmDuration: PT1M
    dcmEntity: Series
    dcmProperty: SendingApplicationEntityTitle=QIDO_JSON
    objectClass: dcmExportRule
    cn: Forward to QIDO-JSON-InstancesOfSeries
    ```` 
- Configure [Qido ETag](https://dcm4chee-arc-cs.readthedocs.io/en/latest/networking/config/archiveNetworkAE.html#dcmqidoetag) as `true` which ensures that archive returns _Last-Modified_ and _ETag_ in responses


Use Case Example : Prefetch [Search Instances of Series of Study](https://petstore.swagger.io/index.html?url=https://raw.githubusercontent.com/dcm4che/dcm4chee-arc-light/master/dcm4chee-arc-ui2/src/swagger/openapi.json#/QIDO-RS/SearchForStudySeriesInstances) 
- On receive of studies, the export rule triggers prefetching of QIDO / WADO by invoking `http[s]://<your-webhook-host>:<webhook-port>/dcm4chee-arc/aets/<ARCHIVE_AET>/rs/studies/[0]/series/[1]/instances`
- When your webhook / cache application receives this request, it shall invoke the same request back to archive replacing `<your-webhook-host>:<webhook-port>` with `<archive-host>:<archive-port>`. You may send one or more of the following request headers : 
   - _If-Modified-Since_
   - _If-Unmodified-Since_
   - _If-Match_
   - _If-None-Match_
- Your webhook / cache application receives [Search Instances of Series of Study](https://petstore.swagger.io/index.html?url=https://raw.githubusercontent.com/dcm4che/dcm4chee-arc-light/master/dcm4chee-arc-ui2/src/swagger/openapi.json#/QIDO-RS/SearchForStudySeriesInstances) response and caches it for faster access by clients.
- Ensure to respond back to initial request sent from archive to webhook / cache application.

**Comment by luxunxiansheng** [comment]
Hi, vrindanayak
      I read the thread carefully.  I wonder if there is simpler and easy way to achive the fast access. Here is a quick thoughts:
 1.  At study or series level,  create a metadata.json file in default. 
 2. Update this json file when a new instance is stored in the archive. 
 3. For study level metadata's WADO request. return this metadata.json file directly. 
 
I am not sure if this is feasible. 

Cheers
L.B.

**Reply by vrindanayak** [reply]
There are different ways to get metadata information from archive which varies based on use cases and application requirements. As mentioned previously, response times of QIDO services are much faster than WADO services because the response is prepared by querying the DB for attributes (in QIDO) vs reading from files (json or DICOM) on filesystem (in WADO) involving IO operations resulting in larger response time compared to QIDO.

Depending upon use case, a client (viewer) may use QIDO services to query the archive or cache, based on its application needs. We have a paying customer (env has millions of studies) which makes use of Apache Traffic Server to prefetch all study / series information on receive of studies by the archive. The viewer systems in turn query ATS Cache or archive directly, depending upon what the end users (doctors / radiologists) request for in the viewer.

Just for info : Wado exporters were redesigned recently #3069 (v5.23.2) to use web applications instead of hard coded URLs - #386 (implemented in 2016), where the web applications configuration (QIDO / WADO) remains independent of exporters configuration - making it simpler to implement / understand both for cache server and archive users. Post this redesigning, pre-fetching QIDO responses - #3215 was implemented

**Reply by luxunxiansheng** [reply]
Suppose not to configure any prefetching  mechanism (QIDO-Prefetch or WADO-prefetch),  is it possible  to get a study-level  metadata ( containing  all  instances' attributes) from DB in one single QIDO or WADO  request ?

**Reply by vrindanayak** [reply]
Yes
- QIDO [Search for Instances of Study](https://petstore.swagger.io/index.html?url=https://raw.githubusercontent.com/dcm4che/dcm4chee-arc-light/master/dcm4chee-arc-ui2/src/swagger/openapi.json#/QIDO-RS/SearchForStudyInstances)
- WADO [Retrieve Study Metadata](https://petstore.swagger.io/index.html?url=https://raw.githubusercontent.com/dcm4che/dcm4chee-arc-light/master/dcm4chee-arc-ui2/src/swagger/openapi.json#/WADO-RS/RetrieveStudyMetadata)

**Reply by luxunxiansheng** [reply]
1.   I will use QIDO for its good performance 
2.  I will query the archive rather than  the cache 
3.  How to  get the result same as a WADO request   "/studies/{study}/series/{series}/metadata"

I had tried to QIDO with /studies/{study}/series/{series}/instances ,  but no detailed meta info of instance contained in the json ( but WADO does have)

Sorry for bothering again and again.

**Reply by vrindanayak** [reply]
QIDO follows the [Search Transaction](http://dicom.nema.org/medical/dicom/current/output/html/part18.html#sect_10.6) whereas WADO follows the [Retrieve Transaction](http://dicom.nema.org/medical/dicom/current/output/html/part18.html#sect_10.4). To get more information in QIDO, you would have to extend the default [Attribute Filters](https://dcm4chee-arc-cs.readthedocs.io/en/latest/networking/config/attributeFilter.html) but doing this (i.e. storing all attributes of instance on receive of studies) is not very helpful as it increases the DB blobs size. 

- What is the average / maximum size of studies (no. of instances) in your env?
- For most clients / viewers, QIDO response should be sufficient for initial loading of data. If they really need some attributes for initial processing which are not available in corresponding QIDO responses, you may extend Attribute Filters with individual attributes accordingly. Else post initial load, the clients / viewers can retrieve metadata as required or you may think of cache servers between clients and archive (with prefetch explained above) for faster retrieval.

**Reply by luxunxiansheng** [reply]
Thanks  a lot. We will try the arribute filter.      
Our avaerage : 200   and max.: 600 or so .   but most of the attributes are not useful. 

BTW:
 I wolud like come back to the pervious topic about prefetch WADO .   From the discussion, it is abvious that a metadata.json file on study or serise level has been generated in the file system.  What not let a wado request 
 "/studies/{study}/series/{series}/metadata" read that json file ?  It is only I/O once and should be faster than to loop each single instance json file .

**Reply by vrindanayak** [reply]
Again, as mentioned previously - using retrieve metadata WADO service to store json file on a storage is an optional use case where the metadata files are passively lying for any third party applications which may need it for post processing. This use case was not intended for WADO services invoked by client / viewer to archive, wherein archive needs to read it from this single metadata file. 
For faster retrievals by client systems / viewers, it is better to simply build / implement a web cache server (ATS) between archive and viewer to reduce the response times considerably by prefetching the metadata (or any other data like images / thumbnails etc) in advance

**Reply by luxunxiansheng** [reply]
Thank you very much !  It is clear now.
